src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SpecializedTasks.cs

```diff
@@ -87,6 +88,19 @@ public static ValueTask<T[]> WhenAll<T>(IEnumerable<ValueTask<T>> tasks)
             }
         }
 
+        [SuppressMessage("Style", "VSTHRD200:Use \"Async\" suffix for async methods", Justification = "Naming is modeled after Task.WhenAll.")]
+        public static async ValueTask<ImmutableArray<TResult>> WhenAll<TResult>(this IReadOnlyCollection<Task<TResult>> tasks)
+        {
+            using var _ = ArrayBuilder<TResult>.GetInstance(tasks.Count, out var result);
+
+            // Explicit cast to IEnumerable<Task> so we call the overload that doesn't allocate an array as the result.
+            await Task.WhenAll((IEnumerable<Task>)tasks).ConfigureAwait(false);
+            foreach (var task in tasks)
+                result.Add(await task.ConfigureAwait(false));
+
+            return result.ToImmutableAndClear();
```

Comment:
ToImmutableAndClear allows for an underlying call to MoveToImmutable.  We likely should always be using it if we use the GetInstance form that takes a count. I'll open another PR on that.